jerry-core/ecma/operations/ecma-objects-general.c

```diff
@@ -592,166 +592,162 @@ ecma_op_general_object_define_own_property (ecma_object_t *obj_p, /**< the objec
-
-  bool is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = true;
-  if (property_desc_p->is_value_defined)
-  {
-    if (!is_current_data_descriptor
-        || !ecma_op_same_value (property_desc_p->value,
-                                ecma_get_named_data_property_value (current_p)))
-    {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
-    }
-  }
-
-  if (property_desc_p->is_writable_defined)
-  {
-    if (!is_current_data_descriptor
-        || property_desc_p->is_writable != ecma_is_property_writable (current_p))
-    {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
-    }
-  }
+  ecma_property_types_t current_property_type = ECMA_PROPERTY_GET_TYPE (current_p);
+  const bool is_current_configurable = ecma_is_property_configurable (current_p);
 
-  if (property_desc_p->is_get_defined)
-  {
-    if (!is_current_accessor_descriptor
-        || property_desc_p->get_p != ecma_get_named_accessor_property_getter (current_p))
-    {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
-    }
-  }
+  JERRY_ASSERT (current_property_type == ECMA_PROPERTY_TYPE_NAMEDDATA
+                || current_property_type == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
 
-  if (property_desc_p->is_set_defined)
+  /* If current is configurable, any change is allowed. */
+  if (!is_current_configurable)
   {
-    if (!is_current_accessor_descriptor
-        || property_desc_p->set_p != ecma_get_named_accessor_property_setter (current_p))
+    // 7.
+    if (property_desc_p->is_configurable
+        || (property_desc_p->is_enumerable_defined
+            && (property_desc_p->is_enumerable != ecma_is_property_enumerable (current_p))))
     {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
+      // a., b.
+      return ecma_reject (is_throw);
     }
   }
 
-  if (property_desc_p->is_enumerable_defined)
+  // 8.
+  if (property_desc_type == ECMA_PROPERTY_TYPE_GENERIC)
   {
-    if (property_desc_p->is_enumerable != ecma_is_property_enumerable (current_p))
-    {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
-    }
+    /* No action required. */
   }
-
-  if (property_desc_p->is_configurable_defined)
+  else if (likely (property_desc_type == current_property_type))
   {
-    if (property_desc_p->is_configurable != ecma_is_property_configurable (current_p))
+    /* If property is configurable, there is no need for checks. */
+    if (unlikely (!is_current_configurable))
     {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
-    }
-  }
+      if (property_desc_type == ECMA_PROPERTY_TYPE_NAMEDDATA)
+      {
+        // 10.
 
-  if (is_every_field_in_desc_also_occurs_in_current_desc_with_same_value)
-  {
-    return ecma_make_simple_value (ECMA_SIMPLE_VALUE_TRUE);
-  }
+        // a.
+        if (!ecma_is_property_writable (current_p))
```

Comment:
I think this could be expressed in a shorter form by writing
```c
        // 10. a. i. & ii.
        if (!ecma_is_property_writable (current_p) 
            && (property_desc_p->is_writable 
                || (property_desc_p->is_value_defined 
                    && !ecma_op_same_value (property_desc_p->value, ecma_get_named_data_property_value (current_p))))
        {
          return ecma_reject (is_throw);
        }
```